### PR TITLE
feat(map_based_prediction): fix object yaw and velocity when necessary

### PR DESCRIPTION
**Pull request by purewater0901**
Signed-off-by: yutaka <purewater0901@gmail.com>

## Description
Currently, map based prediction has a serious issue that cannot handle the wrong yaw angle of the object. In order to fix this problem, I changed the condition and object yaw and velocity when necessary.

https://user-images.githubusercontent.com/43805014/179172812-37ebbfc5-b697-418a-ab87-10f12ff2590a.mp4

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by taikitanaka3** [comment]
[Q]
Just a question.
Does This PR have contradiction to this PR? https://github.com/autowarefoundation/autoware.universe/pull/1348

**Comment by purewater0901** [comment]
@taikitanaka3 No, this PR is independent of https://github.com/autowarefoundation/autoware.universe/pull/1348

File: perception/map_based_prediction/src/map_based_prediction_node.cpp [hunk 1]
```diff
@@ -610,17 +615,33 @@ PredictedObject MapBasedPredictionNode::getPredictedObjectAsCrosswalkUser(
...
     object_pose, object_twist.linear.x * 0.1, object_twist.linear.y * 0.1, 0.0);
-  return tier4_autoware_utils::calcAzimuthAngle(object_pose.position, future_object_pose.position);
+  const auto updated_object_yaw =
+    tier4_autoware_utils::calcAzimuthAngle(object_pose.position, future_object_pose.position);
+
+  object.kinematics.pose_with_covariance.pose.orientation =
+    tier4_autoware_utils::createQuaternionFromYaw(updated_object_yaw);
+
+  // If the object yaw flips, we also flip the velocity
+  if (std::fabs(updated_object_yaw - original_object_yaw) > tier4_autoware_utils::pi / 2.0) {
```
Comment (taikitanaka3) [inline perception/map_based_prediction/src/map_based_prediction_node.cpp:640]: @purewater0901 
calcAzimuthAngle returns -180deg ~180deg using atan2 is it OK without normalizing here?
I think it's necessary

File: perception/map_based_prediction/src/map_based_prediction_node.cpp [hunk 2]
```diff
@@ -610,17 +615,33 @@ PredictedObject MapBasedPredictionNode::getPredictedObjectAsCrosswalkUser(
...
+  // Compute yaw angle from the velocity and position of the object
+  const auto original_object_yaw =
+    tf2::getYaw(object.kinematics.pose_with_covariance.pose.orientation);
   const auto & object_pose = object.kinematics.pose_with_covariance.pose;
   const auto & object_twist = object.kinematics.twist_with_covariance.twist;
   const auto future_object_pose = tier4_autoware_utils::calcOffsetPose(
     object_pose, object_twist.linear.x * 0.1, object_twist.linear.y * 0.1, 0.0);
-  return tier4_autoware_utils::calcAzimuthAngle(object_pose.position, future_object_pose.position);
+  const auto updated_object_yaw =
+    tier4_autoware_utils::calcAzimuthAngle(object_pose.position, future_object_pose.position);
```
Comment (tkimura4) [inline perception/map_based_prediction/src/map_based_prediction_node.cpp:634]: When orientation_availability = SIGN_UNKNOWN, updated_object_yaw should be ±original_object_yaw

File: perception/map_based_prediction/src/map_based_prediction_node.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (purewater0901) [reply perception/map_based_prediction/src/map_based_prediction_node.cpp]: I changed this line, so we don't have this condition any more.

**Review by taikitanaka3 — APPROVED** [review]
LGTM

image

**Pull request merged**